File: mobile/src/main/java/technology/mainthread/apps/gatekeeper/data/RegisterDevices.java
package technology.mainthread.apps.gatekeeper.data;

import android.os.Build;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.iid.FirebaseInstanceId;

import javax.inject.Inject;

import technology.mainthread.apps.gatekeeper.model.firebase.Device;

public class RegisterDevices {

    private final FirebaseAuth firebaseAuth;
    private final FirebaseDatabase firebaseDatabase;
    private final FirebaseInstanceId firebaseInstanceId;

    @Inject
    public RegisterDevices(FirebaseAuth firebaseAuth,
                           FirebaseDatabase firebaseDatabase,
                           FirebaseInstanceId firebaseInstanceId) {
        this.firebaseAuth = firebaseAuth;
        this.firebaseDatabase = firebaseDatabase;
        this.firebaseInstanceId = firebaseInstanceId;
    }

    public void registerDevice() {
        String token = firebaseInstanceId.getToken();
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if (token != null && user != null) {
            Device device = Device.builder()
                    .userId(user.getUid())
                    .deviceName(Build.MANUFACTURER + Build.PRODUCT)
                    .pushToken(token)
                    .build();
            firebaseDatabase.getReference().child("devices").setValue(device);
        }
    }
}

File: mobile/src/main/java/technology/mainthread/apps/gatekeeper/data/RegisterDevices.kt
package technology.mainthread.apps.gatekeeper.data

import android.os.Build
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.iid.FirebaseInstanceId
import technology.mainthread.apps.gatekeeper.model.firebase.Device
import javax.inject.Inject

class RegisterDevices @Inject
constructor(private val firebaseAuth: FirebaseAuth,
            private val firebaseDatabase: FirebaseDatabase,
            private val firebaseInstanceId: FirebaseInstanceId) {

    fun registerDevice() {
        val token = firebaseInstanceId.token
        val user = firebaseAuth.currentUser
        if (token != null && user != null) {
            val device = Device(user.uid, Build.MANUFACTURER + Build.PRODUCT, token)
            firebaseDatabase.reference.child("devices").setValue(device)
        }
    }
}
